### PR TITLE
Update couch_lru to use maps

### PR DESCRIPTION
**Pull request by nickva**
Maps are more ergonomic and usually more performant.

Also add some tests for couch_lru and a few specs with an improved comment indicating what the close function actually doesn't. Previously, it mentions only one of the things: closing oldest idle, however it also bumps busy entries, and clears deleted (or locked) handles from the cache.

This change was originally part of another PR that's now closed: #5165.

Investigation into couch_lru started with issue #5166

### CODE REVIEW

**Review by jaydoane — APPROVED** [review]
Awesome!
```
couch_lru                      :  96%
```

File: src/couch/src/couch_lru.erl [hunk 1]
```diff
@@ -15,34 +15,43 @@
 
 -include("couch_server_int.hrl").
 
+-type cache() :: {any(), #{binary() => pos_integer()}}.
```
[the remark this answers is not in the document]
Comment (nickva) [reply src/couch/src/couch_lru.erl:18]: Oh, that's much better. Will update.

File: src/couch/src/couch_lru.erl [hunk 2]
```diff
@@ -62,5 +71,166 @@ close_int({Lru, DbName, Iter}, {Tree, Dict} = Cache) ->
...
+    ?assertEqual(1, map_size(Map)),
+    ?assertMatch(#{?DB1 := _}, Map),
+    #{?DB1 := Int} = Map,
+    ?assert(is_integer(Int)),
+    ?assert(Int > 0),
+    ?assertEqual([{Int, ?DB1}], gb_trees:to_list(Tree)).
+
+t_insert_duplicate(_) ->
+    % We technically allow this, but is this right? Should we always use update
+    % instead which would reap the old LRU entry
```
[the remark this answers is not in the document]
Comment (nickva) [reply src/couch/src/couch_lru.erl:121]: Oh, I wonder if this is the culprit that causes the bug in https://github.com/apache/couchdb/issues/5166!

[the remark this answers is not in the document]
Comment (nickva) [reply src/couch/src/couch_lru.erl:121]: It's also why we had to add ` We closed this database before processing the update.  Ignore` comment and case in update. That means we somehow ended up with a unsynchronized Tree and Map. Something like:  `tree([Id1 -> db, Id2, -> db,  Id3 -> db])` and a map like `#{db => id3}`

I am thinking we should probably just use update, to ensure the cache tree and map always are in sync with each other at least.

**Pull request merged**